File: firmware/components/protocol/Protocol.cpp
#include "Protocol.h"

#include <cstring>

namespace rfsense::proto {
namespace {

inline void putU16(uint8_t* p, uint16_t v) {
  p[0] = static_cast<uint8_t>(v & 0xFF);
  p[1] = static_cast<uint8_t>((v >> 8) & 0xFF);
}

inline void putU32(uint8_t* p, uint32_t v) {
  p[0] = static_cast<uint8_t>(v & 0xFF);
  p[1] = static_cast<uint8_t>((v >> 8) & 0xFF);
  p[2] = static_cast<uint8_t>((v >> 16) & 0xFF);
  p[3] = static_cast<uint8_t>((v >> 24) & 0xFF);
}

inline void putU64(uint8_t* p, uint64_t v) {
  for (int i = 0; i < 8; ++i) {
    p[i] = static_cast<uint8_t>((v >> (8 * i)) & 0xFF);
  }
}

inline uint16_t getU16(const uint8_t* p) {
  return static_cast<uint16_t>(p[0]) | (static_cast<uint16_t>(p[1]) << 8);
}

inline uint32_t getU32(const uint8_t* p) {
  return static_cast<uint32_t>(p[0]) | (static_cast<uint32_t>(p[1]) << 8) |
         (static_cast<uint32_t>(p[2]) << 16) | (static_cast<uint32_t>(p[3]) << 24);
}

inline uint64_t getU64(const uint8_t* p) {
  uint64_t v = 0;
  for (int i = 0; i < 8; ++i) {
    v |= static_cast<uint64_t>(p[i]) << (8 * i);
  }
  return v;
}

}  // namespace

uint32_t crc32Update(uint32_t crc, const uint8_t* data, std::size_t len) {
  crc = ~crc;
  for (std::size_t i = 0; i < len; ++i) {
    crc ^= data[i];
    for (int b = 0; b < 8; ++b) {
      const uint32_t mask = -(crc & 1u);
      crc = (crc >> 1) ^ (0xEDB88320u & mask);
    }
  }
  return ~crc;
}

uint32_t crc32(const uint8_t* data, std::size_t len) { return crc32Update(0u, data, len); }

DatagramBuilder::DatagramBuilder(uint8_t* buffer, std::size_t capacity)
    : buf_(buffer), cap_(capacity) {}

void DatagramBuilder::begin(const DatagramHeader& header) {
  frameCount_ = 0;
  offset_ = 0;
  begun_ = false;
  if (cap_ < kDatagramHeaderSize + kCrcSize) {
    return;  // buffer too small to ever hold a valid datagram
  }
  buf_[0] = kMagic0;
  buf_[1] = kMagic1;
  buf_[2] = kMagic2;
  buf_[3] = kMagic3;
  buf_[4] = header.protocolVersion;
  buf_[5] = header.flags;
  buf_[6] = static_cast<uint8_t>(header.captureMode);
  buf_[7] = 0;  // reserved0
  putU32(buf_ + 8, header.deviceId);
  putU32(buf_ + 12, header.bootId);
  putU32(buf_ + 16, header.packetSeq);
  putU32(buf_ + 20, header.batchSeq);
  putU16(buf_ + 24, 0);  // frameCount placeholder
  putU16(buf_ + 26, 0);  // payloadLen placeholder
  putU32(buf_ + 28, 0);  // reserved1
  offset_ = kDatagramHeaderSize;
  begun_ = true;
}

bool DatagramBuilder::canFit(uint16_t csiLen) const {
  if (!begun_) {
    return false;
  }
  const std::size_t need = kFrameFixedSize + csiLen;
  const std::size_t projected = offset_ + need + kCrcSize;
  return projected <= cap_ && projected <= kMaxDatagramSize;
}

bool DatagramBuilder::addFrame(const FrameHeader& frame, const uint8_t* csi) {
  if (!canFit(frame.csiLen)) {
    return false;
  }
  uint8_t* p = buf_ + offset_;
  putU32(p + 0, frame.frameSeq);
  putU64(p + 4, frame.timestampUs);
  putU32(p + 12, frame.pingSeq);
  p[16] = static_cast<uint8_t>(frame.rssi);
  p[17] = static_cast<uint8_t>(frame.noiseFloor);
  p[18] = frame.channel;
  p[19] = frame.secondaryChannel;
  p[20] = frame.bandwidth;
  p[21] = frame.phyMode;
  p[22] = frame.rate;
  p[23] = frame.firstWordInvalid;
  putU16(p + 24, frame.linkId);
  putU16(p + 26, frame.csiLen);
  if (frame.csiLen > 0 && csi != nullptr) {
    std::memcpy(p + kFrameFixedSize, csi, frame.csiLen);
  }
  offset_ += kFrameFixedSize + frame.csiLen;
  ++frameCount_;
  return true;
}

std::size_t DatagramBuilder::finalize() {
  if (!begun_) {
    return 0;
  }
  const uint16_t payloadLen = static_cast<uint16_t>(offset_ - kDatagramHeaderSize);
  putU16(buf_ + 24, frameCount_);
  putU16(buf_ + 26, payloadLen);
  const uint32_t crc = crc32(buf_, offset_);
  putU32(buf_ + offset_, crc);
  offset_ += kCrcSize;
  return offset_;
}

ParseStatus parseHeader(const uint8_t* buf, std::size_t len, DatagramHeader& out) {
  if (len < kDatagramHeaderSize + kCrcSize) {
    return {false, "datagram too short"};
  }
  if (buf[0] != kMagic0 || buf[1] != kMagic1 || buf[2] != kMagic2 || buf[3] != kMagic3) {
    return {false, "bad magic"};
  }
  out.protocolVersion = buf[4];
  if (out.protocolVersion != kProtocolVersion) {
    return {false, "unsupported protocol version"};
  }
  out.flags = buf[5];
  out.captureMode = static_cast<CaptureMode>(buf[6]);
  out.deviceId = getU32(buf + 8);
  out.bootId = getU32(buf + 12);
  out.packetSeq = getU32(buf + 16);
  out.batchSeq = getU32(buf + 20);
  out.frameCount = getU16(buf + 24);
  out.payloadLen = getU16(buf + 26);
  if (kDatagramHeaderSize + out.payloadLen + kCrcSize != len) {
    return {false, "payloadLen mismatch"};
  }
  return {true, nullptr};
}

bool verifyCrc(const uint8_t* buf, std::size_t len) {
  if (len < kDatagramHeaderSize + kCrcSize) {
    return false;
  }
  const std::size_t bodyLen = len - kCrcSize;
  const uint32_t expected = crc32(buf, bodyLen);
  const uint32_t actual = getU32(buf + bodyLen);
  return expected == actual;
}

bool parseFrame(const uint8_t* buf, std::size_t len, std::size_t& byteOffset, FrameHeader& out,
                const uint8_t** csiOut) {
  if (byteOffset + kFrameFixedSize > len) {
    return false;
  }
  const uint8_t* p = buf + byteOffset;
  out.frameSeq = getU32(p + 0);
  out.timestampUs = getU64(p + 4);
  out.pingSeq = getU32(p + 12);
  out.rssi = static_cast<int8_t>(p[16]);
  out.noiseFloor = static_cast<int8_t>(p[17]);
  out.channel = p[18];
  out.secondaryChannel = p[19];
  out.bandwidth = p[20];
  out.phyMode = p[21];
  out.rate = p[22];
  out.firstWordInvalid = p[23];
  out.linkId = getU16(p + 24);
  out.csiLen = getU16(p + 26);
  if (byteOffset + kFrameFixedSize + out.csiLen > len) {
    return false;
  }
  if (csiOut != nullptr) {
    *csiOut = p + kFrameFixedSize;
  }
  byteOffset += kFrameFixedSize + out.csiLen;
  return true;
}

}  // namespace rfsense::proto

File: tools/collector/src/protocol.ts
// Binary CSI streaming protocol v1 decoder.
//
// This mirrors firmware/components/protocol/Protocol.cpp byte-for-byte. Any change to the wire
// format must be made in both places and in docs/csi-protocol.md. Raw CSI bytes are never
// altered here -- they are surfaced exactly as the device sent them.

export const MAGIC = Uint8Array.from([0x52, 0x46, 0x43, 0x53]); // "RFCS"
export const PROTOCOL_VERSION = 1;
export const DATAGRAM_HEADER_SIZE = 32;
export const FRAME_FIXED_SIZE = 28;
export const CRC_SIZE = 4;
export const MAX_DATAGRAM_SIZE = 1400;
export const PING_SEQ_NONE = 0xffffffff;
export const FLAG_MAINTENANCE = 0x01;

export enum CaptureMode {
  Controlled = 0,
  Normal = 1,
  Passive = 2,
}

export interface DatagramHeader {
  protocolVersion: number;
  flags: number;
  captureMode: number;
  deviceId: number;
  bootId: number;
  packetSeq: number;
  batchSeq: number;
  frameCount: number;
  payloadLen: number;
}

export interface FrameRecord {
  frameSeq: number;
  timestampUs: number; // microseconds since device boot (fits in a double for centuries)
  pingSeq: number; // PING_SEQ_NONE when the frame is not ping-correlated
  rssi: number;
  noiseFloor: number;
  channel: number;
  secondaryChannel: number;
  bandwidth: number;
  phyMode: number;
  rate: number;
  firstWordInvalid: number;
  linkId: number;
  csiLen: number;
  csi: Buffer; // raw signed I/Q bytes, copied out verbatim
}

export interface Datagram {
  header: DatagramHeader;
  frames: FrameRecord[];
}

export type ParseResult = { ok: true; datagram: Datagram } | { ok: false; error: string };

// IEEE 802.3 CRC32 (reflected, poly 0xEDB88320) -- the same polynomial the firmware uses.
const CRC_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c >>> 0;
  }
  return table;
})();

export function crc32(data: Uint8Array, start = 0, end = data.length): number {
  let crc = 0xffffffff;
  for (let i = start; i < end; i++) {
    crc = CRC_TABLE[(crc ^ data[i]!) & 0xff]! ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

export function parseDatagram(buf: Buffer): ParseResult {
  if (buf.length < DATAGRAM_HEADER_SIZE + CRC_SIZE) {
    return { ok: false, error: 'datagram too short' };
  }
  if (buf[0] !== MAGIC[0] || buf[1] !== MAGIC[1] || buf[2] !== MAGIC[2] || buf[3] !== MAGIC[3]) {
    return { ok: false, error: 'bad magic' };
  }
  const protocolVersion = buf[4]!;
  if (protocolVersion !== PROTOCOL_VERSION) {
    return { ok: false, error: `unsupported protocol version ${protocolVersion}` };
  }
  const header: DatagramHeader = {
    protocolVersion,
    flags: buf[5]!,
    captureMode: buf[6]!,
    deviceId: buf.readUInt32LE(8),
    bootId: buf.readUInt32LE(12),
    packetSeq: buf.readUInt32LE(16),
    batchSeq: buf.readUInt32LE(20),
    frameCount: buf.readUInt16LE(24),
    payloadLen: buf.readUInt16LE(26),
  };
  if (DATAGRAM_HEADER_SIZE + header.payloadLen + CRC_SIZE !== buf.length) {
    return { ok: false, error: 'payloadLen mismatch' };
  }

  const bodyLen = buf.length - CRC_SIZE;
  const expected = crc32(buf, 0, bodyLen);
  const actual = buf.readUInt32LE(bodyLen);
  if (expected !== actual) {
    return { ok: false, error: 'crc mismatch' };
  }

  const frames: FrameRecord[] = [];
  let offset = DATAGRAM_HEADER_SIZE;
  for (let i = 0; i < header.frameCount; i++) {
    if (offset + FRAME_FIXED_SIZE > bodyLen) {
      return { ok: false, error: 'frame header out of bounds' };
    }
    const csiLen = buf.readUInt16LE(offset + 26);
    const csiStart = offset + FRAME_FIXED_SIZE;
    if (csiStart + csiLen > bodyLen) {
      return { ok: false, error: 'csi out of bounds' };
    }
    frames.push({
      frameSeq: buf.readUInt32LE(offset),
      timestampUs: Number(buf.readBigUInt64LE(offset + 4)),
      pingSeq: buf.readUInt32LE(offset + 12),
      rssi: buf.readInt8(offset + 16),
      noiseFloor: buf.readInt8(offset + 17),
      channel: buf[offset + 18]!,
      secondaryChannel: buf[offset + 19]!,
      bandwidth: buf[offset + 20]!,
      phyMode: buf[offset + 21]!,
      rate: buf[offset + 22]!,
      firstWordInvalid: buf[offset + 23]!,
      linkId: buf.readUInt16LE(offset + 24),
      csiLen,
      csi: Buffer.from(buf.subarray(csiStart, csiStart + csiLen)),
    });
    offset = csiStart + csiLen;
  }
  if (offset !== bodyLen) {
    return { ok: false, error: 'trailing bytes after frames' };
  }
  return { ok: true, datagram: { header, frames } };
}
